### ISSUE TITLE
WIP: Law checking & extending IOEffect typeclasses

### ISSUE THREAD

**Issue by benhutchison** [atnos-org/eff#134]
This issue is to track some experimental [hacking](https://github.com/benhutchison/eff-cats/blob/laws/cats/src/main/scala/org/atnos/eff/addon/cats/effect/IOEffect.scala) Ive been doing around

- Supporting the `ConcurrentEffect` typeclass from cats.effect 0.10 in [IOEffect](https://github.com/benhutchison/eff-cats/blob/bf1468df577ebe108518dfe5bd46cf875ae5b670/cats/src/main/scala/org/atnos/eff/addon/cats/effect/IOEffect.scala#L109). This is a very speculative, "follow-the-types" implementation and I don't feel confident of its correctness or even sanity - any feedback most welcome. There's some [gnarly transient mutable state]() to work around `runIO`s inablity to thread non-`Unit` values down the chain, inspired by @etorreborre hint on [gitter](https://gitter.im/atnos-org/eff?at=5a1c0438232e79134dd91699) last Nov

- Law checking of the [Async instance](https://github.com/atnos-org/eff/blob/master/cats/src/main/scala/org/atnos/eff/addon/cats/effect/IOEffect.scala#L80) and this new `ConcurrentEffect` using Discipline. On the `laws` branch, `cats/test` now runs law checks. One finding was that the existing Async fails 10 law checks, once around [stack safety](https://gist.github.com/benhutchison/192f1e380a3b958464477fc607b04929#file-sbt_output-txt-L228) and nine times around [seemingly not suspending thrown Throwables](https://gist.github.com/benhutchison/192f1e380a3b958464477fc607b04929#file-sbt_output-txt-L27).

The story behind why I got interested in this area is that I will teach a second [Eff workshop](http://lambdajam.yowconference.com.au/proposal/?id=6016) at Lambdajam this year. I want to update my material and try to teach whatever is contemporary "best-practice" as at 2018, and I think using cats.effect.IO with Eff would be a nice touch.

However that said, Ive already got the `IO` effect working fine inside Eff stack, so getting these typeclasses law-checked/working is perhaps a bad habit of looking down passing rabbit-holes :P 

CC @alexandru

**Comment by benhutchison** [comment]
OK, Ive not made enough progress today and have hit the limits of what my knowledge & timebox can deliver. Closing this issue, reopen if I learn something that can unblock.

Concurrent law checks enabled, shows [many are failing](https://gist.github.com/benhutchison/36ebfcaa159ffe0185422d404855cbe2).

Earlier issues with Async laws were resolved with a better `Eq` check. Its still failing stack safety, related to  https://github.com/typelevel/cats-effect/issues/42

**Comment by etorreborre** [comment]
Stack safety issues might be related to the fact that [I couldn't implement it properly for Eff](https://github.com/atnos-org/eff/blob/master/shared/src/main/scala/org/atnos/eff/Eff.scala#L244). I think a proper implementation. I think the correct implementation would go along the lines of [FreeApplicative](https://github.com/typelevel/cats/blob/664cb624d4dcede7c968e82d2ff0c4e5981df065/free/src/main/scala/cats/free/FreeApplicative.scala#L59) which is quite involved.

**Comment by alexandru** [comment]
The stack safety laws are pretty severe — being required for both left-associated and right-associated `flatMap` chains, and also for `map` in recursive data structures too.

Sometimes I wonder if maybe they are too severe, but then there are instances in which we really need it (e.g. with `map` at least I've got the use-case in `Iterant` where if `map` isn't stack-safe, most operations could blow up with stack overflow errors for strictly initialized streams).

I don't know if it helps, not familiar with your implementation, but make sure to checkout the shenanigans we now have in:

- `IndexedStateT`: https://github.com/typelevel/cats/pull/2187
- `Kleisli`: https://github.com/typelevel/cats/pull/2185

Also note that the laws need validation. `Concurrent` is very fresh for example, so when a law is failing try to understand why it fails and if that's *fair* or not — some laws might have assumptions about the execution model in them that we aren't aware of. It happened before, coming to light after I implemented `Async` for `Iterant` for example, which led to patching the laws to be more relaxed.

It's basically an ongoing process with those laws, so feedback is appreciated.
